apps/fortuna/src/command/run.rs

```diff
@@ -121,38 +122,69 @@ pub async fn run_keeper(
 
 pub async fn run(opts: &RunOptions) -> Result<()> {
     let config = Config::load(&opts.config.config)?;
+    let provider_config = ProviderConfig::load(&opts.provider_config.provider_config)?;
     let private_key = opts.load_private_key()?;
     let secret = opts.randomness.load_secret()?;
     let (tx_exit, rx_exit) = watch::channel(false);
 
     let mut chains: HashMap<ChainId, BlockchainState> = HashMap::new();
     for (chain_id, chain_config) in &config.chains {
         let contract = Arc::new(PythContract::from_config(&chain_config)?);
+        let provider_config = provider_config.get_chain_config(chain_id)?;
+        let provider_commitments = &provider_config.commitments;
         let provider_info = contract.get_provider_info(opts.provider).call().await?;
+        let latest_metadata =
+            bincode::deserialize::<CommitmentMetadata>(&provider_info.commitment_metadata)?;
 
-        // Reconstruct the hash chain based on the metadata and check that it matches the on-chain commitment.
-        // TODO: we should instantiate the state here with multiple hash chains.
-        // This approach works fine as long as we haven't rotated the commitment (i.e., all user requests
-        // are for the most recent chain).
         // TODO: we may want to load the hash chain in a lazy/fault-tolerant way. If there are many blockchains,
         // then it's more likely that some RPC fails. We should tolerate these faults and generate the hash chain
         // later when a user request comes in for that chain.
-        let metadata =
-            bincode::deserialize::<CommitmentMetadata>(&provider_info.commitment_metadata)?;
 
-        let hash_chain = PebbleHashChain::from_config(
+        // Reconstruct the hash chain for old registrations
+        let mut offsets = provider_commitments
+            .iter()
+            .map(|x| {
+                x.original_commitment_sequence_number
+                    .try_into()
+                    .map_err(|_| anyhow!("{}: error converting u64 offset to usize", chain_id))
+            })
+            .collect::<Result<Vec<usize>>>()?;
+
+        let mut hash_chains = provider_commitments
+            .iter()
+            .map(|x| {
+                let metadata =
```

Comment:
these creation logics are duplicated both for offsets and hash_chains. You can add the latest_commitment to the end of provider_commitments array and do these only once.